## Facies classification using plurality voting (e.g. multiclass majority voting)


#### Contest entry by: <a href="https://github.com/mycarta">Matteo Niccoli</a> and <a href="https://github.com/dahlmb">Mark Dahl</a>

####  [Original contest notebook](../Facies_classification.ipynb) by Brendon Hall, [Enthought](https://www.enthought.com/)



<a rel="license" href="http://creativecommons.org/licenses/by/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by/4.0/88x31.png" /></a><br /><span xmlns:dct="http://purl.org/dc/terms/" property="dct:title">The code and ideas in this notebook,</span> by <span xmlns:cc="http://creativecommons.org/ns#" property="cc:attributionName">Matteo Niccoli and Mark Dahl,</span> are licensed under a <a rel="license" href="http://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution 4.0 International License</a>.

In this notebook we will attempt to predict facies from well log data using machine learnig classifiers.  The dataset comes from a class exercise from The University of Kansas on [Neural Networks and Fuzzy Systems](http://www.people.ku.edu/~gbohling/EECS833/).  This exercise is based on a consortium project to use machine learning techniques to create a reservoir model of the largest gas fields in North America, the Hugoton and Panoma Fields. For more info on the origin of the data, see [Bohling and Dubois (2003)](http://www.kgs.ku.edu/PRS/publication/2003/ofr2003-50.pdf) and [Dubois et al. (2007)](http://dx.doi.org/10.1016/j.cageo.2006.08.011). 

The dataset consists of log data from nine wells that have been labeled with a facies type based on observation of core. We will use this log data to train a support vector machine to classify facies types. 

### The plan

We will created three classifiers with pretuned parameters:
- best SVM in the competition (by our team's SVM submission) 
- best Random Forest in the competition (form the leading submission, by gccrowther) 
- multilayer perceptron (from previous notebooks, not submitted)

We will then try to predict the facies using a plurality voting approach (plurality voting = multi-class majority voting).

From the [scikit-learn website](http://scikit-learn.org/stable/modules/ensemble.html#voting-classifier): "The idea behind the voting classifier implementation is to combine conceptually different machine learning classifiers and use a majority vote or the average predicted probabilities (soft vote) to predict the class labels. Such a classifier can be useful for a set of equally well performing model in order to balance out their individual weaknesses".


### Exploring the dataset

First, we will examine the data set we will use to train the classifier.

In [1]:
%matplotlib inline
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.colors as colors
from mpl_toolkits.axes_grid1 import make_axes_locatable

import pandas as pd
from pandas import set_option
set_option("display.max_rows", 10)
pd.options.mode.chained_assignment = None

from sklearn import preprocessing
from sklearn.metrics import f1_score, accuracy_score, make_scorer
from sklearn.model_selection import LeaveOneGroupOut

In [2]:
filename = 'facies_vectors.csv'
training_data = pd.read_csv(filename)
training_data

Facies Formation        Well Name   Depth      GR  ILD_log10  DeltaPHI  \
0          3     A1 SH        SHRIMPLIN  2793.0  77.450      0.664     9.900   
1          3     A1 SH        SHRIMPLIN  2793.5  78.260      0.661    14.200   
2          3     A1 SH        SHRIMPLIN  2794.0  79.050      0.658    14.800   
3          3     A1 SH        SHRIMPLIN  2794.5  86.100      0.655    13.900   
4          3     A1 SH        SHRIMPLIN  2795.0  74.580      0.647    13.500   
...      ...       ...              ...     ...     ...        ...       ...   
4144       5      C LM  CHURCHMAN BIBLE  3120.5  46.719      0.947     1.828   
4145       5      C LM  CHURCHMAN BIBLE  3121.0  44.563      0.953     2.241   
4146       5      C LM  CHURCHMAN BIBLE  3121.5  49.719      0.964     2.925   
4147       5      C LM  CHURCHMAN BIBLE  3122.0  51.469      0.965     3.083   
4148       5      C LM  CHURCHMAN BIBLE  3122.5  50.031      0.970     2.609   

       PHIND     PE  NM_M  RELPOS  
0     11.915  4.600     1   1.000  
1     12.565  4.100     1   0.979  
2     13.050  3.600     1   0.957  
3     13.115  3.500     1   0.936  
4     13.300  3.400     1   0.915  
...      ...    ...   ...     ...  
4144   7.254  3.617     2   0.685  
4145   8.013  3.344     2   0.677  
4146   8.013  3.190     2   0.669  
4147   7.708  3.152     2   0.661  
4148   6.668  3.295     2   0.653  

[4149 rows x 11 columns]

This data is from the Council Grove gas reservoir in Southwest Kansas.  The Panoma Council Grove Field is predominantly a carbonate gas reservoir encompassing 2700 square miles in Southwestern Kansas.  This dataset is from nine wells (with 4149 examples), consisting of a set of seven predictor variables and a rock facies (class) for each example vector and validation (test) data (830 examples from two wells) having the same seven predictor variables in the feature vector.  Facies are based on examination of cores from nine wells taken vertically at half-foot intervals. Predictor variables include five from wireline log measurements and two geologic constraining variables that are derived from geologic knowledge. These are essentially continuous variables sampled at a half-foot sample rate. 

The seven predictor variables are:
* Five wire line log curves include [gamma ray](http://petrowiki.org/Gamma_ray_logs) (GR), [resistivity logging](http://petrowiki.org/Resistivity_and_spontaneous_%28SP%29_logging) (ILD_log10),
[photoelectric effect](http://www.glossary.oilfield.slb.com/en/Terms/p/photoelectric_effect.aspx) (PE), [neutron-density porosity difference and average neutron-density porosity](http://petrowiki.org/Neutron_porosity_logs) (DeltaPHI and PHIND). Note, some wells do not have PE.
* Two geologic constraining variables: nonmarine-marine indicator (NM_M) and relative position (RELPOS)

The nine discrete facies (classes of rocks) are: 
1. Nonmarine sandstone
2. Nonmarine coarse siltstone 
3. Nonmarine fine siltstone 
4. Marine siltstone and shale 
5. Mudstone (limestone)
6. Wackestone (limestone)
7. Dolomite
8. Packstone-grainstone (limestone)
9. Phylloid-algal bafflestone (limestone)

These facies aren't discrete, and gradually blend into one another. Some have neighboring facies that are rather close.  Mislabeling within these neighboring facies can be expected to occur.  The following table lists the facies, their abbreviated labels and their approximate neighbors.

Facies |Label| Adjacent Facies
:---: | :---: |:--:
1 |SS| 2
2 |CSiS| 1,3
3 |FSiS| 2
4 |SiSh| 5
5 |MS| 4,6
6 |WS| 5,7
7 |D| 6,8
8 |PS| 6,7,9
9 |BS| 7,8

Let's clean up this dataset.  The 'Well Name' and 'Formation' columns can be turned into a categorical data type.  

In [3]:
training_data['Well Name'] = training_data['Well Name'].astype('category')
training_data['Formation'] = training_data['Formation'].astype('category')
training_data['Well Name'].unique()

[SHRIMPLIN, ALEXANDER D, SHANKLE, LUKE G U, KIMZEY A, CROSS H CATTLE, NOLAN, Recruit F9, NEWBY, CHURCHMAN BIBLE]
Categories (10, object): [SHRIMPLIN, ALEXANDER D, SHANKLE, LUKE G U, ..., NOLAN, Recruit F9, NEWBY, CHURCHMAN BIBLE]

These are the names of the 10 training wells in the Council Grove reservoir.  Data has been recruited into pseudo-well 'Recruit F9' to better represent facies 9, the Phylloid-algal bafflestone. 

Before we plot the well data, let's define a color map so the facies are represented by consistent color in all the plots in this tutorial.  We also create the abbreviated facies labels, and add those to the `facies_vectors` dataframe.

In [4]:
# 1=sandstone  2=c_siltstone   3=f_siltstone # 4=marine_silt_shale 
#5=mudstone 6=wackestone 7=dolomite 8=packstone 9=bafflestone
facies_colors = ['#F4D03F', '#F5B041', '#DC7633','#A569BD',
       '#000000', '#000080', '#2E86C1', '#AED6F1', '#196F3D']

facies_labels = ['SS', 'CSiS', 'FSiS', 'SiSh', 'MS',
                 'WS', 'D','PS', 'BS']
#facies_color_map is a dictionary that maps facies labels
#to their respective colors
facies_color_map = {}
for ind, label in enumerate(facies_labels):
    facies_color_map[label] = facies_colors[ind]

def label_facies(row, labels):
    return labels[ row['Facies'] -1]
    
training_data.loc[:,'FaciesLabels'] = training_data.apply(lambda row: label_facies(row, facies_labels), axis=1)
training_data.describe()

Facies        Depth           GR    ILD_log10     DeltaPHI  \
count  4149.000000  4149.000000  4149.000000  4149.000000  4149.000000   
mean      4.503254  2906.867438    64.933985     0.659566     4.402484   
std       2.474324   133.300164    30.302530     0.252703     5.274947   
min       1.000000  2573.500000    10.149000    -0.025949   -21.832000   
25%       2.000000  2821.500000    44.730000     0.498000     1.600000   
50%       4.000000  2932.500000    64.990000     0.639000     4.300000   
75%       6.000000  3007.000000    79.438000     0.822000     7.500000   
max       9.000000  3138.000000   361.150000     1.800000    19.312000   

             PHIND           PE         NM_M       RELPOS  
count  4149.000000  3232.000000  4149.000000  4149.000000  
mean     13.201066     3.725014     1.518438     0.521852  
std       7.132846     0.896152     0.499720     0.286644  
min       0.550000     0.200000     1.000000     0.000000  
25%       8.500000     3.100000     1.000000     0.277000  
50%      12.020000     3.551500     2.000000     0.528000  
75%      16.050000     4.300000     2.000000     0.769000  
max      84.400000     8.094000     2.000000     1.000000

This is a quick view of the statistical distribution of the input variables.  Looking at the `count` values, most values have 4149 valid values except for `PE`, which has 3232.  We will drop the feature vectors that don't have a valid `PE` entry.

In [5]:
PE_mask = training_data['PE'].notnull().values
training_data = training_data[PE_mask]

In [6]:
training_data.describe()

Facies        Depth           GR    ILD_log10     DeltaPHI  \
count  3232.000000  3232.000000  3232.000000  3232.000000  3232.000000   
mean      4.422030  2875.824567    66.135769     0.642719     3.559642   
std       2.504243   131.006274    30.854826     0.241845     5.228948   
min       1.000000  2573.500000    13.250000    -0.025949   -21.832000   
25%       2.000000  2791.000000    46.918750     0.492750     1.163750   
50%       4.000000  2893.500000    65.721500     0.624437     3.500000   
75%       6.000000  2980.000000    79.626250     0.812735     6.432500   
max       9.000000  3122.500000   361.150000     1.480000    18.600000   

             PHIND           PE         NM_M       RELPOS  
count  3232.000000  3232.000000  3232.000000  3232.000000  
mean     13.483213     3.725014     1.498453     0.520287  
std       7.698980     0.896152     0.500075     0.286792  
min       0.550000     0.200000     1.000000     0.010000  
25%       8.346750     3.100000     1.000000     0.273000  
50%      12.150000     3.551500     1.000000     0.526000  
75%      16.453750     4.300000     2.000000     0.767250  
max      84.400000     8.094000     2.000000     1.000000

Now we extract just the feature variables we need to perform the classification.  The predictor variables are the five log values and two geologic constraining variables, **and we are also using depth**. We also get a vector of the facies labels that correspond to each feature vector.

In [7]:
y = training_data['Facies'].values
print y[25:40]
print np.shape(y)

[3 3 2 2 2 2 2 2 3 3 3 3 3 3 3]
(3232,)


In [8]:
X = training_data.drop(['Formation', 'Well Name','Facies','FaciesLabels'], axis=1)
print np.shape(X)
X.describe(percentiles=[.05, .25, .50, .75, .95])

(3232, 8)


Depth           GR    ILD_log10     DeltaPHI        PHIND  \
count  3232.000000  3232.000000  3232.000000  3232.000000  3232.000000   
mean   2875.824567    66.135769     0.642719     3.559642    13.483213   
std     131.006274    30.854826     0.241845     5.228948     7.698980   
min    2573.500000    13.250000    -0.025949   -21.832000     0.550000   
5%     2632.775000    23.491000     0.237299    -5.600000     4.800000   
25%    2791.000000    46.918750     0.492750     1.163750     8.346750   
50%    2893.500000    65.721500     0.624437     3.500000    12.150000   
75%    2980.000000    79.626250     0.812735     6.432500    16.453750   
95%    3061.500000   106.268000     1.045606    12.000000    27.787400   
max    3122.500000   361.150000     1.480000    18.600000    84.400000   

                PE         NM_M       RELPOS  
count  3232.000000  3232.000000  3232.000000  
mean      3.725014     1.498453     0.520287  
std       0.896152     0.500075     0.286792  
min       0.200000     1.000000     0.010000  
5%        2.523300     1.000000     0.070000  
25%       3.100000     1.000000     0.273000  
50%       3.551500     1.000000     0.526000  
75%       4.300000     2.000000     0.767250  
95%       5.369550     2.000000     0.962000  
max       8.094000     2.000000     1.000000

In [9]:
scaler = preprocessing.StandardScaler().fit(X)
X = scaler.transform(X)

### Make performance scorers
Used to evaluate performance.

In [10]:
Fscorer = make_scorer(f1_score, average = 'micro')

### Pre-tuned SVM classifier classifier

In [11]:
from sklearn import svm
SVC_classifier = svm.SVC(C = 100, cache_size=2400, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma=0.01, kernel='rbf',
  max_iter=-1, probability=True, random_state=1, shrinking=True,
  tol=0.001, verbose=False)

### Pre-tuned multi-layer perceptron classifier

In [12]:
from sklearn.neural_network import MLPClassifier
mlp_classifier = MLPClassifier(activation='logistic', alpha=0.01, batch_size='auto',
       beta_1=0.9, beta_2=0.999, early_stopping=False, epsilon=1e-08,
       hidden_layer_sizes=(100,), learning_rate='adaptive',
       learning_rate_init=0.001, max_iter=1000, momentum=0.9,
       nesterovs_momentum=True, power_t=0.5, random_state=1, shuffle=True,
       solver='adam', tol=0.0001, validation_fraction=0.1, verbose=False,
       warm_start=False)

### Pre-tuned extra trees
This is the classifier from the leading submission, by gccrowther.

In [13]:
from sklearn.pipeline import make_pipeline
from sklearn.feature_selection import VarianceThreshold
from sklearn.ensemble import ExtraTreesClassifier

ET_classifier = make_pipeline(
    VarianceThreshold(threshold=0.37),
    ExtraTreesClassifier(criterion="entropy", max_features=0.71,
                         n_estimators=500, random_state=49)
    )

### Plurality voting classifier (multi-class majority voting)

In [14]:
from sklearn.ensemble import VotingClassifier

In [15]:
eclf = VotingClassifier(estimators=[
        ('SVC', SVC_classifier), ('MLP', mlp_classifier), ('ET', ET_classifier)], 
                        voting='soft', weights=[0.3,0.2,0.5])

### Confusion matrix and average test F1 score

Let's see how we do with predicting the actual facies, by looking at a confusion matrix. We do this by keeping the parameters from the previous section, but creating a brand new classifier. So when we fit the data, it won't have seen it before.

In [16]:
eclf.fit(X,y)
pred = eclf.predict(X)

from sklearn.metrics import confusion_matrix
from classification_utilities import display_cm, display_adj_cm

conf = confusion_matrix(pred, y)
display_cm(conf, facies_labels, display_metrics=True, hide_zeros=True)

     Pred    SS  CSiS  FSiS  SiSh    MS    WS     D    PS    BS Total
     True
       SS   259                                                   259
     CSiS         738                                             738
     FSiS               615                                       615
     SiSh                     184                                 184
       MS                           217                           217
       WS                                 462                     462
        D                                        98                98
       PS                                             498         498
       BS                                                   161   161

Precision  1.00  1.00  1.00  1.00  1.00  1.00  1.00  1.00  1.00  1.00
   Recall  1.00  1.00  1.00  1.00  1.00  1.00  1.00  1.00  1.00  1.00
       F1  1.00  1.00  1.00  1.00  1.00  1.00  1.00  1.00  1.00  1.00


In [17]:
f1_ens = []

wells = training_data["Well Name"].values
logo = LeaveOneGroupOut()

eclf_cv = VotingClassifier(estimators=[
        ('SVC', SVC_classifier), ('MLP', mlp_classifier), ('ET', ET_classifier)], 
        voting='soft', weights=[2,3,1])

for train, test in logo.split(X, y, groups=wells):
    well_name = wells[test[0]]
    eclf_cv.fit(X[train], y[train])
    pred_cv = eclf_cv.predict(X[test])
    sc = f1_score(y[test], pred_cv, labels = np.arange(10), average = 'micro')
    print("{:>20s}  {:.3f}".format(well_name, sc))
    f1_ens.append(sc)
    
print "-Average leave-one-well-out F1 Score: %6f" % (sum(f1_ens)/(1.0*(len(f1_ens))))

     CHURCHMAN BIBLE  0.554
      CROSS H CATTLE  0.363
            LUKE G U  0.438
               NEWBY  0.430
               NOLAN  0.482
          Recruit F9  0.809
             SHANKLE  0.512
           SHRIMPLIN  0.601
-Average leave-one-well-out F1 Score: 0.523695


### Predicting, and saving facies for blind wells

In [18]:
blind = pd.read_csv('validation_data_nofacies.csv') 
X_blind = np.array(blind.drop(['Formation', 'Well Name'], axis=1)) 
X_blind = scaler.transform(X_blind) 
y_pred = eclf_cv.fit(X, y).predict(X_blind) 
blind['Facies'] = y_pred